### PR TITLE
SNOW-1300193: Add error message when trying to read from local directory

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1300193

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   This change adds an error message for when DataframeReader is passed a non-stage path. The SQL error for this case can be confusing to users so having a more direct error is desirable.

### CODE REVIEW

File: src/snowflake/snowpark/dataframe_reader.py [hunk 1]
```diff
@@ -59,6 +59,14 @@
 }
 
 
+def _validate_stage_path(path: str) -> str:
+    if not path.strip("\"'").startswith("@"):
+        raise ValueError(
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/dataframe_reader.py:64]: do we also need to take care of snow urls `snow://`?

Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/dataframe_reader.py:64]: I'm unfamiliar with those. Is there good documentation somewhere?

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/dataframe_reader.py [hunk 2]
```diff
@@ -411,6 +421,7 @@ def csv(self, path: str) -> DataFrame:
         Returns:
             a :class:`DataFrame` that is set up to load data from the specified CSV file(s) in a Snowflake stage.
         """
+        path = _validate_stage_path(path)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe_reader.py:424]: I think we need a changelog for this behavior that the path is eagerly validated

File: tests/integ/scala/test_dataframe_reader_suite.py [hunk 3]
```diff
@@ -594,12 +594,10 @@ def test_read_csv_with_format_type_options(session, mode, local_testing_mode):
...
-    df2 = get_reader(session, mode).schema(user_schema).csv(test_file_csv_colon)
-    with pytest.raises(SnowparkSQLException) as ex_info:
-        df2.collect()
-    assert (
-        "SQL compilation error" if not local_testing_mode else "Invalid stage"
-    ) in str(ex_info)
+    with pytest.raises(
+        ValueError, match="DataFrameReader can only read files from stage locations."
+    ):
+        get_reader(session, mode).schema(user_schema).csv(test_file_csv_colon)
```
Comment (sfc-gh-aling) [inline tests/integ/scala/test_dataframe_reader_suite.py:600]: do you think this is a error BCR that previously customer gets SnowparkSQLException but now it's ValueError?

File: src/snowflake/snowpark/dataframe_reader.py [hunk 4]
```diff
@@ -706,6 +717,7 @@ def _read_semi_structured_file(self, path: str, format: str) -> DataFrame:
 
         if self._user_schema:
             raise ValueError(f"Read {format} does not support user schema")
+        path = _validate_stage_path(path)
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/dataframe_reader.py:720]: do we need this change for other format like json, avro, parquet methods

Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/dataframe_reader.py:720]: Those methods call _read_semi_structured_file. I've updated tests to show that they are also affected.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**